src/audio/kpb.c

```diff
@@ -171,45 +179,56 @@ static int kpb_prepare(struct comp_dev *dev)
 
 	cd->no_of_clients = 0;
 
-	/* allocate history_buffer/s */
-#if KPB_NO_OF_HISTORY_BUFFERS == 2
+	/* allocate history buffer/s */
+	while (hb_size > 0 && i < ARRAY_SIZE(hb_mcp)) {
+		ptr = rballoc(RZONE_RUNTIME, hb_mcp[i], ca_size);
+
+		if (ptr) {
+			history_buffer->start_addr = ptr;
+			history_buffer->end_addr = ptr + ca_size;
+			history_buffer->w_ptr = ptr;
+			history_buffer->r_ptr = ptr;
+			history_buffer->state = KPB_BUFFER_FREE;
+			hb_size = hb_size - ca_size;
+			history_buffer->next = &cd->history_buffer;
+
+			/* Do we need another buffer? */
+			if (hb_size > 0) {
+				new_hb = rzalloc(RZONE_RUNTIME,
+						 SOF_MEM_CAPS_RAM,
+						 COMP_SIZE(struct hb));
+				history_buffer->next = new_hb;
+				new_hb->state = KPB_BUFFER_OFF;
+				new_hb->next = history_buffer;
+				history_buffer = new_hb;
+				ca_size = hb_size;
+				i++;
+			}
 
-	cd->his_buf_hp.sta_addr = rballoc(RZONE_RUNTIME, SOF_MEM_CAPS_RAM,
-					  KPB_MAX_BUFFER_SIZE - LPSRAM_SIZE);
-	cd->his_buf_lp.sta_addr = rballoc(RZONE_RUNTIME, SOF_MEM_CAPS_LP,
-					  LPSRAM_SIZE);
-
-	if (!cd->his_buf_hp.sta_addr || !cd->his_buf_lp.sta_addr) {
-		trace_kpb_error("Failed to allocate space for "
-				"KPB bufefrs");
-		return -ENOMEM;
+		} else {
+			/* we've failed to allocate ca_size of that hb_mcp
+			 * let's try again with some smaller size.
+			 * NOTE! If we decrement by some small value,
+			 * the allocation will take significant time.
+			 * However, bigger values like HEAP_HP_BUFFER_BLOCK_SIZE
+			 * will result in lower accuracy of allocation.
+			 */
+			ca_size = ca_size - KPB_ALLOCATION_STEP;
+			if (ca_size <= 0) {
+				ca_size = hb_size;
+				i++;
+			}
+			continue;
+		}
 	}
 
-	cd->his_buf_hp.end_addr = cd->his_buf_hp.sta_addr +
-	(KPB_MAX_BUFFER_SIZE - LPSRAM_SIZE);
-
-	cd->his_buf_lp.end_addr = cd->his_buf_lp.sta_addr +
-	LPSRAM_SIZE;
-
-#elif KPB_NO_OF_HISTORY_BUFFERS == 1
-
-	cd->his_buf_hp.sta_addr = rballoc(RZONE_RUNTIME, SOF_MEM_CAPS_RAM,
-					  KPB_MAX_BUFFER_SIZE);
-
-	if (!cd->his_buf_hp.sta_addr) {
+	/* have we succeed in allocation of at least one buffer? */
+	if (!cd->history_buffer.start_addr) {
```

Comment:
@mrajwa what if the first history buffer allocation succeeded and the second one failed?
isnt if (hb_size) a better check?